### ISSUE TITLE
Add support to use TPM 2.0 Preboot Event log on Keylime

### ISSUE THREAD

**Issue by maugustosilva** [keylime/keylime#491]
Adding all the relevant parties here: @puiterwijk  @mpeters @lukehinds @kaifeng 

# Motivation
We have a scenario with literally hundreds of different types of nodes, and simply "getting the contents of the boot aggregate (PCR 10) from a list of well-known values" has shown itself to be unfeasible from an operational standpoint (just to give some perspective: our nodes are netbooted, and the value of every PCR1 is different as it encodes the MAC address of the NIC). Our experiments with the "TPM2 Preboot Event log" demonstrated that, with a "recent enough" Kernel (e.g. 5.4.X), and a "recent enough" version of `tpm2-tools`, we can access the contents of the aforementioned log to re-create (or debug) the values of PCRs [0-9],14 without requiring a pre-measurement (i.e., a "golden value"). Therefore, I would like to propose we add this capability to Keylime.

# Summary
In a nutshell all we need is for the `agent` to be able to read the (binary) contents of `/sys/kernel/security/tpm0/binary_bios_measurements` and send it back to the `verifier`, where it would be "processed" in a manner similar to what we already do with the IMA runtime measurements.

# Estimated impact
This is not dissimilar to what is already done with IMA in the sense that: a) it is an optional capability, which should not interfere with the "regular" operation of KL and b) adds, to each interaction `verifier`<->`agent` interaction, no more than tens of KB (which is the total size of the "TPM2 Preboot Event log"). While we do expect very little impact on KeyLime's scalability by adding this capability, it is important to remember that we are constantly testing KL in a configuration with 5K nodes (with IMA), and can provide experimental evidence to back this hypothesis.  

# Plan of action
If you all agree with this change, I would like to propose a 3-phase implementation (i.e., roughly 3 individual PRs):

1) Retrieve boot logs in the `agent`, but do nothing with them in the `verifier` (`verifier` just receives log and ignores it), with new `keylime_tenant` CLI options (that was a request from @mpeters) to signal that this "tpm2 boot event log" should be read. This should be very similar to what we do currently with IMA, a new CLI option, hard-coded PCR (10 in case of IMA, [0-9], 14 in case of boot log) and hard-coded paths on the `/sys` (`/sys/kernel/security/ima/ascii_runtime_measurements` in case of IMA, `/sys/kernel/security/tpm0/binary_bios_measurements` in case of boot log) filesystem.

2) Adds the ability to authenticate the boot log (i.e., just replays the log and obtains the same values for the aforementioned PCRs), but does not perform any semantic analysis of it (this more intelligent semantic analysis will performed later by a "policy engine"). This requires a particular version of the Intel TPM2 tool kit (has to fail gracefully if that is not available)

3) Introduces the concept of a "policy engine", which will be able to compare the contents of the extracted - by the `agent` - "TPM2 reboot event log" against a "template". This "template" could be understood as something similar to an IMA "allowlist", and therefore should also be stored on the `verifier` database. Please note this template would vary from "empty" (meaning, "accept everything as long as the values of the PCRs [0,9],14 match the contents of the log") to "exact" (meaning, "we are providing the exact boot log for each node, and they simply need to match bit-by-bit"). Our particular use case will be with some form of "template", where we can just match against a list of diverse values (e.g., we don't care about the specific MAC address of the NIC or the specific firmware release of a card... they just need to be within acceptable bounds). Finally, it is important to note that the "policy engine", will read "policies" (in the form of python files) which are proprietary and not open-sourced.

**Comment by lukehinds** [comment]
@puiterwijk

**Comment by kaifeng** [comment]
I have seen description of event log in the TCG spec, it seems it was stored somewhere in the bios but didn't figure out how to read it out, this proposal seems quite related with it.

**Comment by maugustosilva** [comment]
@kaifeng yes, this is correct. The log is accessible via `/sys/kernel/security/tpm0/binary_bios_measurements`

**Comment by galmasi** [comment]
Related question -- this came up in a discussion yesterday, and we did not have a ready answer to it. Is *any* keylime user putting PCR0-9 + 14 values to use in any way today? our current implementation excludes these PCRs from the "TPM policy" when measured boot is turned on.

**Comment by galmasi** [comment]
Note that in this issue we are specifically not addressing the problem of measured boot on TPM1 devices. We had a discussion with @mpeters and the general feeling seems to be that "TPM1 is on the way out".
